### PR TITLE
[Merged by Bors] - Directory restructure

### LINKED ISSUE #1427: Two copies of validator_definitions.yml
## Description

I'm seeing two copies of `validator_definitions.yml`, one in `~/.altona/` and the other in `~/.altona/validators` (`~/.altona` is my validator dir). I don't have time to investigate now, but it looks like there's something strange going on

## Version

Lighthouse 0.1.2-2ede9caa

### LINKED ISSUE #1487: Directory restructure for multiple networks
## Description

Currently, we have the default directory in lighthouse as `~/.lighthouse/`. Unless we explicitly provide the `--datadir` flag, the beacon/validator/account commands use this directory as the default base directory. The folder has the following structure 

```
.lighthouse
├── beacon
├── secrets
├── validators
└── wallets
```
Currently, we have only a single network that people care about (medalla), so it's fine to have the beacon and key related subdirectories at the top level. 

But as we move closer towards mainnet, we will have multiple networks that we would want to run and we could see issues with overwriting of one chain's data/keys with another.

## Proposal 

```
.lighthouse
└── medalla
    ├── beacon
    ├── secrets
    ├── validators
    └── wallets
└── mainnet
    ├── beacon
    ├── secrets
    ├── validators
    └── wallets
└── custom
    ├──...
```
Here, we have a top level directory for every named testnet we can pass in the `--testnet` option and a custom directory for handling networks with a  custom `--testnet-dir`. 
Having multiple unnamed testnets from the `--testnet-dir` can lead to overwriting of files in the `custom` directory.

The disadvantage I see with this proposal is that this isn't backwards compatible for users currently running medalla. 
So we would need to have a mechanism for moving the current top level folders that are being used in medalla inside `~/.lighthouse/medalla` the first time a user runs the beacon/validator/account binaries. Not sure if this would be an issue with permissions somehow.

Keen to hear your thoughts/suggestions/alternatives :)

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Closes #1487
Closes #1427

## Proposed Changes

Directory restructure in accordance with #1487. Also has temporary migration code to move the old directories into new structure.
Also extracts all default directory names and utility functions into a `directory` crate to avoid repetitio.

## Additional info

~Since `validator_definition.yaml` stores absolute paths, users will have to manually change the keystore paths or delete the file to get the validators picked up by the vc.~. `validator_definition.yaml` is migrated as well from the default directories.

### CODE REVIEW

**Comment by AgeManning** [comment]
I'm keen to get this in.

This might also mess with the docker. Might need to check the docker file and scripts directory and potentially update the compose (in a separate repo). 

Once updated to master, lets get a few approvals and merge this guy in :)

**Review by michaelsproul — CHANGES_REQUESTED** [review]
I'm keen for this change too! Our datadirs have been a mess for too long.

It looks like there are a few more hardcoded instances of `.lighthouse` in libp2p and `lcli` that should be refactored to use the new style.

```
$ git grep '"\.lighthouse' '**.rs'
beacon_node/eth2_libp2p/src/config.rs:75:        network_dir.push(".lighthouse");
common/directory/src/lib.rs:6:pub const DEFAULT_ROOT_DIR: &str = ".lighthouse";
consensus/types/src/test_utils/builders/testing_beacon_state_builder.rs:17:        .map(|home| (home.join(".lighthouse")))
lcli/src/eth1_genesis.rs:23:                .map(|home| home.join(".lighthouse").join("testnet"))
lcli/src/interop_genesis.rs:34:                .map(|home| home.join(".lighthouse").join("testnet"))
lcli/src/new_testnet.rs:13:        PathBuf::from(".lighthouse/testnet"),
```

The `keypairs_path` function in `test_utils` is unused and can be deleted outright.

I dream of all our datadirs being consistent and centrally managed :relieved:

File: lighthouse/src/main.rs
Comment (pawanjay176) [reply lighthouse/src/main.rs:210]: @michaelsproul even if we move the `validator_definitions.yaml`, the keystore paths and keystore password paths within the file will point to wrong directories since they are absolute paths. 
If we delete the definitions file, then the vc will repopulate it but users would loose their manually entered passwords from the `voting_keystore_password` key in the yaml.

So should we be changing the contents of the definitions file and then move the file to reflect the new paths as well as part of the migration?

Comment (michaelsproul) [reply lighthouse/src/main.rs:210]: Oh damn, that's a good point. We could attempt an auto-migration but maybe it should be non-destructive (i.e. don't delete the original files -- or move them out of the way).

We should probably also ship the release with a big warning about possibly having to manually intervene to fix things if the automatic migration fails.

Comment (michaelsproul) [reply lighthouse/src/main.rs:210]: Summoning @paulhauner and @AgeManning

Comment (pawanjay176) [reply lighthouse/src/main.rs:214]: Added a hacky way to migrate the `validator_definitions.yaml` in https://github.com/sigp/lighthouse/pull/1532/commits/ad22178ed6d4591f3fbeef4d42e3438b8904afd2

Comment (AgeManning) [reply lighthouse/src/main.rs:214]: Yeah. I'm of the opinion that software is still experimental. We can do huge breaking releases. 

Try and hack the upgrade but warn people. Best to rip the band-aid off early imo.

**Comment by pawanjay176** [comment]
This is ready for review now. I have tested the migration and it seems to work fine with binaries and docker (except for a few extra directories in the docker version. Will fix that in the lighthouse-docker repo).

Apart from the directory restructure and migration, I have also made a few opinionated changes in the account manager and validator binary cli parsing. 

The `--datadir` flag is described as `Root data directory for lighthouse keys and databases.`. However, this flag wasn't used at all in the `account_manager` binary. Also in the `validator` binary, we had to pass a path to a directory which contained the keys directly instead of the root data directory (equivalent to `$HOME/.lighthouse/validators` instead of `$HOME/.lighthouse`). 


This PR changes `--datadir` in all binaries(bn, vc and am) to mean the base directory which would contain the `beacon`, `validators`, `wallets` and `secrets` folders.

In the `validator` binary, users can also provide separate directories for `validators` and `secrets` by using the `--validators-dir` and `--secrets-dir` flags together. Similarly in the `account_manager` binary for providing `validators`, 'wallets' and `secrets` directories.

File: testing/node_test_rig/src/lib.rs
Comment (michaelsproul) [inline testing/node_test_rig/src/lib.rs:173]: We could rename `ValidatorFiles::datadir` to `validator_dir` too

File: validator_client/src/config.rs
Comment (michaelsproul) [inline validator_client/src/config.rs:88]: double space between "data" and "does"

**Review by michaelsproul — CHANGES_REQUESTED** [review]
This is looking really good.

I tested the migration on a temporary local node and it worked well! My main Medalla validators use a custom datadir, so the migration won't happen automatically there, but I'm confident I'll be able to tweak it manually. We'll have to make it clear in our announcement all the necessary steps to migrate a custom datadir, and the expected locations of files.

Due to the high risk of misplacing the slashing protection database during the changeover, I think we should bring back a flag to disable auto-registration of validators in the slashing protection database. If a user mistakenly starts their validator client with the slashing DB in the wrong location, the VC will silently recreate `slashing_protection.sqlite`, and register all the validators from `validator_definitions.yml` with this new (empty!) database. It would be really easy to get slashed.

The `--disable-auto-discover` flag from #1357 is not sufficient protection, for the reasons highlighted above, so I think we really need a flag like `--disable-slashing-registration` to skip this function call:

https://github.com/sigp/lighthouse/blob/9c9176c1d1c24de87e297aa4a90f9353dfab0405/validator_client/src/lib.rs#L203-L205

Long-term I think `disable-slashing-registration` should be the default, and the only time new validators should get enabled in the slashing protection database is when they're created (I've made issue #1584 to track this)

File: common/directory/src/lib.rs
Comment (pawanjay176) [reply common/directory/src/lib.rs:19]: Yep, agree. I initially had the function to fill the `Default` impls of the various configs, but ended up using it in other places myself :sweat_smile: Will remove this :)

File: beacon_node/eth2_libp2p/src/config.rs
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/config.rs:74]: Yep, I just assumed that the `network_dir` default would always get overrided. But I'll add that and get rid of `get_default_base_dir` as well.

**Comment by pawanjay176** [comment]
> We'll have to make it clear in our announcement all the necessary steps to migrate a custom datadir, and the expected locations of files.

Not sure I understand. There's no testnet specific folders within a datadir. So if you are using a custom datadir, then there's no need for migration. The location of files and folders would be exactly same as before.

The only difference would be the path of the datadir in the vc would have changed.
Where earlier it would be
```sh
lighthouse --datadir custom_dir/ bn
lighthouse --datadir custom_dir/validators vc
```

Now it would be 
```sh
lighthouse --datadir custom_dir/ bn
lighthouse --datadir custom_dir/ vc
```

Fair point regarding the slashing protection database. Will add the `--disable-slashing-registration` flag.

**Comment by michaelsproul** [comment]
> The only difference would be the path of the datadir in the vc would have changed.

Yeah, this is the issue. On my node I'm running with `lighthouse vc --datadir custom_dir`, which weirdly means that my datadir is laid out like this:

```
.medalla/
  beacon/
  secrets/
  slashing_protection.sqlite
  slashing_protection.sqlite-journal
  validator_definitions.yml
  validators/
    0xa0...
  wallets/
```

I must have moved some things around by hand to get `validators` to be different from the `lighthouse vc` datadir. I think the only reason it works is that the validator key loader searches the directory recursively :flushed:

I would definitely get slashed if I carelessly switched to the new `--datadir .medalla` flag, because my `slashing_protection.sqlite` is not in `.medalla/validators` where it should be.

I think the `--disable-slashing-registration` flag will be a nice safeguard for people like me :laughing: . Having thought about the name some more I think we should call it something more positive sounding like `--strict-slashing-protection`. Disabling something related to slashing doesn't immediately give the impression of being safer

**Comment by pawanjay176** [comment]
Ohh I didn't consider that the slashing protection database and validator_definitions could be outside the validators directory. Makes sense now! 

Agree with the `--strict-slashing-protection` name!

**Review by michaelsproul — APPROVED** [review]
I'm happy with this, but I'll leave the r+ to @paulhauner or @AgeManning, who have authority to make a release and a Discord announcement.

I think we should cut a new v0.2.9 release, and in the announcement:

* Highlight the datadir change, and that `~/.lighthouse` will be automatically migrated to `~/.lighthouse/medalla`
* Recommend that anyone using a custom dir takes extra care when upgrading and runs with `--strict-slashing-protection` on their validator client to ensure their slashing protection database remains in use after the upgrade. Users who have trouble should move their `slashing_protection.sqlite` and `validator_defintions.yml` into their `<datadir>/validators` directory, and point their `--datadir` flag at `<datadir>`, or use the new `--validators-dir` flag to point to `<datadir>/validators`.

**Comment by paulhauner** [comment]
I'll go through this in detail in the coming day or two, sorry for not getting across it early.

**Comment by yorickdowne** [comment]
Why does `validator_definition.yml` store absolute paths? This makes it difficult to move validator data to a new `--datadir` location. Is there a concern about storing relative paths, relative to `--datadir`?

**Comment by pawanjay176** [comment]
> Why does `validator_definition.yml` store absolute paths?

The main reason is that the `validators` and `secrets` directories need not necessarily have the same parent to derive a common relative path.
We can specify them separately using the `validators-dir` and `secrets-dir` flags.

**Comment by paulhauner** [comment]
> This makes it difficult to move validator data to a new --datadir location

Supporting relative paths might be something we can add in the future. It's just that absolute paths are much easier to reason about.

> You can delete the validator_definitions.yml folder after moving it to a separate location and the validator_client will regenerate it with the new paths for the keys and secrets.

The validator definitions file can store keystore passwords, they will need to be re-entered if you do this.

File: validator_client/src/lib.rs
Comment (paulhauner) [inline validator_client/src/lib.rs:70]: I think we should change `datadir` -> `validator_dir`.

File: validator_client/src/cli.rs
Comment (paulhauner) [inline validator_client/src/cli.rs:29]: Should we also add this `conflicts_with` to `--secrets-dir`?

**Review by paulhauner — CHANGES_REQUESTED** [review]
I really like this new directory structure, good job at dealing with all the intricacies. 

I only had nitpicks on general structure, but I have some more substantial thoughts about migration (see below)

## Migration

I'm of the opinion that we should abandon the auto-migration stuff for the following reasons:

1. It's quite complex and subject to weird edge cases (which I've listed below).
1. Since we're going to release this with `v0.3.0` which requires a resync, it's not really valuable to migrate the `beacon/` directory anymore.

If we drop the auto-migrate, we should be able to make a fairly simple accompanying announcement that if you have an existing `~/lighthouse` directory, you'll need to move all the files into a `medalla`/`spadina` directory.

https://github.com/sigp/lighthouse-docker might have some issues with this, but I think we can solve it by simply adding a `--datadir` flag to the bn/vc start scripts.

### Migration edge cases

1. `ValidatorDefinitions::migrate` assumes that all keystores live in the same directory in `~/.lighthouse`, which is not always the case.
1. If the user has an empty `~/lighthouse/validators` directory then `lighthouse bn` will error with `Failed to open validator_definitions.yml` and do a partial migration. Next time `lighthouse bn` is run, it won't detect the partial migration and will start successfully.
1. It's possible that a user has an existing Medalla DB but first runs this code using `--testnet spadina`. In that case, they will have a `~/.lighthouse/spadina` directory with a Medalla DB in it.
1. Another variation on the previous point is that the user runs with `--testnet medalla` and gets a migration into the `~/.lighthouse/medalla` testnet (this is correct). Then, since the `beacon/` dir isn't deleted, if they run with `--testnet spadina` they'll get that same `beacon/` dir migrated into the `~/.lighthouse/spadina` directory (this is invalid).

## Summary

So, moving forward, I suggest that we remove the auto-migration stuff and keep the rest of it as-is.

**Comment by michaelsproul** [comment]
+1 for removing the migration now that we're planning a resyncable v0.3 release

**Comment by paulhauner** [comment]
IMO it's worth trying a `git rebase v0.3.0-staging`, it might be easier than a standard `git merge v0.3.0-staging`, but I'm not sure.

**Comment by pawanjay176** [comment]
Agree with removing the auto migration stuff. Thanks for pointing it out :+1:

File: account_manager/src/common.rs
Comment (michaelsproul) [inline account_manager/src/common.rs:34]: Might be nice to keep this comment

**Review by michaelsproul — APPROVED** [review]
Approved!

Keen to get this merged into v0.3.0-staging so I can base #1544 (or similar) on it

**Review by paulhauner — APPROVED** [review]
Awesome!

**Pull request closed without merging**
